boofuzz/socket_connection.py

```diff
@@ -148,7 +149,10 @@ def open(self):
             if self._udp_broadcast:
                 self._sock.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, True)
         elif self.proto == "raw-l2":
-            self._sock = socket.socket(socket.AF_PACKET, socket.SOCK_RAW)
+            raw_proto = self.ethernet_proto if self.ethernet_proto else 0
+            self._sock = socket.socket(socket.AF_PACKET, socket.SOCK_RAW, socket.htons(raw_proto))
+            if self.bind:
+                self._sock.bind(self.bind)
```

Comment:
We can either use `socket.bind()` while opening or `socket.sendto()` later. Using both is redundant.

However, I haven't quite figured out what the right format of the tuple inside bind() would be for AF_PACKET.
The [docs](https://docs.python.org/3/library/socket.html#socket-families) at AF_PACKET might help here. Maybe something like `bind(self.host, raw_proto)` works. Also I don't think we should reuse the `bind` argument which was meant to be used for UDP receptions.
Have you fully tested this all? Currently, I don't really have the time to do so.

Right now boofuzz does not support receiving on raw sockets at all. While sending, there is no use for filtering by ethertype or MAC. If we manage to get a proper raw packet filter working, we could start to implement the receiving function. Otherwise I don't really see a point in these changes here or am I missing something?

Reply:
Nope, you've assessed the situation perfectly. My current recieving setup is a bit... janky to say the least and I need to work on that. I plan to add support for recieving on raw sockets as well, however this may be taking another week or two depending on my workload.

Personally, I'd stay with `socket.bind()` over `socket.sendto()`, especially with receiving in mind. The Tuple inside bind() -- for the AF_PACKET socket -- selects an interface and a protocol (which, for some reason, you have to supply to `socket()` as well). This, to me, seems to stem from some python weirdness, in raw C you would have to supply the protocol in the sockaddr struct to bind as well. Reading `man 7 packet` helped me getting a better understanding of that. For example, to send and capture a multitude of 802.11 frames on my wifi adapter ath0 I'd use `("ath0", ETH_P_ALL)` with `ETH_P_ALL = 0x0003` taken from [<net/if_ether.h>](https://github.com/torvalds/linux/blob/master/include/uapi/linux/if_ether.h)

As for testing: I haven't tested the patch per se apart from ensuring that the socket opens like i'd expect it to. However, in a little script to test stuff opening a socket like this works like expected, so I see no reason why it shouldn't do what I expect here.

As for the bind argument: I don't really see your concern here, apart that I forgot to update the documentation. The bind argument behaves with either protocol exactly how I would expect it to and one is able to guess from the name what happens behind the scenes.

Reply:
Sounds good, no hurry.

Agreed, we should use `socket.bind()`, but consistently.
As for the tuple, `self._sock.bind(self.host, raw_proto)` should work fine then. Why would we want to use the `bind` argument for that? Wouldn't the information in that argument be redundant with `host` and `ethernet_proto`?

Ok, we don't really need any special testing in the current state, but once the receiving is implemented, we have to verify that the packet filter works as intended.